### ISSUE TITLE
[tensorflow-cc] Build error

### ISSUE THREAD

**Issue by MimTaha** [microsoft/vcpkg#30582]
**Host Environment**

- OS: [Windows 10]
- Compiler: vs2022

**To Reproduce**

`./vcpkg install tensorflow-cc`

**Failure logs**

- vcpkg-tool version: 2022-12-14-7ae0d8527fb488fde10a89c2813802dc9b03b6f9
- vcpkg-scripts version: d1e2ff0c3 2023-03-27 (4 days ago).

**Additional context**

error: building tensorflow-cc:x64-windows failed with: BUILD_FAILED

**Comment by Adela0814** [comment]
```
Downloading https://mirrors.sjtug.sjtu.edu.cn/msys2/mingw/x86_64/mingw-w64-x86_64-python-numpy-1.20.3-1-any.pkg.tar.zst
error: Failed to download from mirror set
error: https://repo.msys2.org/mingw/x86_64/mingw-w64-x86_64-python-numpy-1.20.3-1-any.pkg.tar.zst: failed: status code 404
```

**Comment by dg0yt** [comment]
Can you test the branch from #29067? It would also update the python-numpy package. But tensorflow-cc isn't tested in vcpkg CI for pull requests, and in the full CI the port fails due to lack of bazel.

**Comment by e-mokhov** [comment]
After #29067  merged to the master I tried to install tensorflow-cc, same configuration, I got an error:

> CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: C:/MyProj/vcpkg/downloads/tools/msys2/3a900090e97ce03b/mingw64/bin/python3.exe C:/MyProj/vcpkg/buildtrees/tensorflow-cc/x64-windows-dbg/configure.py --workspace C:/MyProj/vcpkg/buildtrees/tensorflow-cc/x64-windows-dbg
    Working Directory: C:/MyProj/vcpkg/buildtrees/tensorflow-cc/x64-windows-dbg
    Error code: 1   
    See logs for more information:
      C:\MyProj\vcpkg\buildtrees\tensorflow-cc\config-x64-windows-dbg-out.log
      

> Call Stack (most recent call first):
  installed/x64-windows/share/tensorflow-common/tensorflow-common.cmake:187 (vcpkg_execute_required_process)
  ports/tensorflow-cc/portfile.cmake:9 (include)
  scripts/ports.cmake:147 (include)
 
 config-x64-windows-dbg-out.log contains error about bazel:
> Cannot find bazel. Please install bazel/bazelisk.

But bazel and bazelisk are installed and configured, when I run the python directly I got following:
> ./downloads/tools/msys2/3a900090e97ce03b/mingw64/bin/python3.exe ./buildtrees/tensorflow-cc/x64-windows-dbg/configure.py --workspace ./buildtrees/tensorflow-cc/x64-windows-dbg
You have bazel 6.2.0 installed.

**Comment by dg0yt** [comment]
The tensorflow scripts try to use bazel from port `vcpkg-tool-bazel` (which is not version 6.2.0):
~~~cmake
find_program(BAZEL bazel PATHS "${CURRENT_HOST_INSTALLED_DIR}/tools" REQUIRED)
~~~
However, the installed windows executable lacks the `.exe` suffix. It is even visible in the CI file lists:
~~~
vcpkg-tool-bazel:x64-windows:/share/vcpkg-tool-bazel/vcpkg.spdx.json
vcpkg-tool-bazel:x64-windows:/share/vcpkg-tool-bazel/vcpkg_abi_info.txt
vcpkg-tool-bazel:x64-windows:/tools/bazel
~~~
Yes, bazel is the windows program:
~~~
$ ( cd packages/vcpkg-tool-bazel_x64-mingw-static/tools ; file bazel )
bazel: PE32+ executable (console) x86-64, for MS Windows
~~~
This goes unnoticed in vcpkg CI because the tool isn't used - tensorflow is skipped for x64-windows targets.

I try to cover `vcpkg-tool-bazel` in a test port now, and will fix the installation in another step once CI complains about the current state: https://github.com/microsoft/vcpkg/pull/31704

**Comment by dg0yt** [comment]
@e-mokhov Can you test #31704? (I have no way to build tensorflow*:x64-windows.)

**Comment by e-mokhov** [comment]
@dg0yt I tested #31704  (in your branch bazel)
I got the same error!
Do I need to downgrade bazel? Use 5.2.0 Instead of 6.2.0? 
Or do something else?

**Comment by dg0yt** [comment]
@e-mokhov Did you actually re-install port `vcpkg-tool-bazel` before trying `tensorflow-cc`?

https://github.com/microsoft/vcpkg/pull/31704#issuecomment-1568010317 reports another issue, but I would be glad if you could get further than `Cannot find bazel.`

**Comment by e-mokhov** [comment]
@dg0yt , you are right. 
I reinstalled `vcpkg-tool-bazel` and get some similar error in
 `build-x64-windows-dbg-err.log`:
> Extracting Bazel installation...
Starting local Bazel server and connecting to it...
ERROR: Expected value after --jobs
/usr/bin/bash: line 1: 12: command not found
/usr/bin/bash: line 1: --compilation_mode=dbg: command not found

**Comment by e-mokhov** [comment]
@dg0yt  I'm trying to test your recommendation https://github.com/microsoft/vcpkg/pull/31704#issuecomment-1568278373, 
what package I need to reinstall to activate your change?

**Comment by dg0yt** [comment]
@e-mokhov You can edit `ports/tensorflow-common/tensorflow-common.cmake` and then reinstall port `tensorflow-common`.

BTW I can confirm that your orignal problem is a baseline regression in vcpkg: It also appears in the nightly vcpkg CI build logs.

**Comment by dg0yt** [comment]
@e-mokhov  Thanks for the feedback. I did't expect the windows port to be so broken...
- [x] Fix msys numpy
- [x] bazel executable
- [ ] bazel option passing

 I think it is time to let vcpkg CI at least temporarily check tensorflow-cc in my PR.

**Comment by michaelaeriksen** [comment]
Is there an ETA on when this will be resolved.

**Comment by dg0yt** [comment]
@michaelaeriksen I'm just a community member.

I don't have resources to debug tensorflow windows builds locally. I can only inspect failure logs from vcpkg CI. 

Can you tell me when did it work the last time?

The windows part hasn't seen maintenance from the vcpkg team in the past months. According to the CI baseline notes, "tensorflow does not support VS2022". I don't know if this is true. What I have found is that
- I can make `tensorflow` builds succeed in CI for x64-windows and x64-windows-static-md.
- I cannot make `tensorflow-cc` builds succeed in CI for any windows triplet.
  It fails in finally producing the DLL, from a generated DEF file and unknown inputs. There is an 18 MB log file of "unresolved external symbol" errors. (Cf. https://dev.azure.com/vcpkg/public/_build/results?buildId=90431&view=artifacts&type=publishedArtifacts.)
  This step is taken even for static library linkage.

Details:
- The tensorflow ports' scripting code isn't really following good practices with regard to portability, scoping etc.
- The tensorflow ports have received many updates with testing of windows triplets in vcpkg CI. However, there are significant windows-only code paths.
- Bazel is a build system for very specific goals which don't integrate easily with vcpkg.
- Tensorflow add it is own level of isolation and scripting on top of that. In particular: a specific processing of DEF files creating Windows DLLs.
- vcpkg adds its own patching on top of that: Port `tensorflow-cc` for windows is configured to build a static lib only - despite tensorflow_cc normally being a DLL.
- Many build details are not visible from the logs, due to use of response files and due to helper compiling scripts into executable.
- The use msys2 and its python adds some complications, in particular with regard to different path conventions.

The next steps:
- Make port `tensorflow` build in CI for some windows tripets.
- With `tensorflow` building for linux, osx, windows, try to update `bazel` from 5.2.0 to 5.4.1.
  This is a LTS version and should be compatible. 
  However, Tensorflow 2.10 reference version is Bazel 5.1.1.

From that point, this could be done in separate efforts:
- Refactor tensorflow script code.
- Update tensorflow to 2.10.1.
  Or update to 2.12. However, GPU support on native windows is no longer available since 2.11.
- Fix the `tensorflow-cc` issues.
  Maybe it needs a community member with MSVC 2019 to contribute to this effort.

**Comment by martinelles** [comment]
Thanks for the effort and explanation @dg0yt
